### ISSUE TITLE
Plugin and/or passed-callable mechanisms for customizing `Inventory.suggest()` scoring

### ISSUE THREAD

**Issue by bskinn** [bskinn/sphobjinv#207]
*To be completed. Related discussion is [here](https://github.com/sopel-irc/sopel-rtfm/commit/6d2de967fb14a287da70d5461f75ca40fa6ae6b5#commitcomment-55538884).

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by eirrgang** [comment]
To some extent, the plug-in mechanism and the plug-in selection may be decoupled.

How would a command-line user select an alternative scoring mechanism? Unless either all detected scorers would be used, or some heuristic would select a scoring mechanism based on the query, then presumably a command line option would name an alternate scoring function, right?

Once the alternative is named by the user, you could scan the `entry_points` for the `sphobjinv` namespace for the named implementation. But you could also just use the convention of looking for `toolname-pluginname`, and use `importlib` to try to import a specifically-named thing from `sphobjinv-<name>`. These are not mutually exclusive, and could be combined with other ways of registering plugins, like environment variables or config files.

Special plugin names is easy and common. Sphinx seems to encourage `sphinx-tool` or `sphinxcontrib-tool` package names, but doesn't rely on the convention internally, as far as I can tell. Pytest [encourages](https://docs.pytest.org/en/latest/how-to/plugins.html#using-plugins) the `pytest-name` convention, but doesn't use it for [run-time discovery](https://docs.pytest.org/en/latest/how-to/writing_plugins.html#plugin-discovery-order-at-tool-startup) ([preferring entry points](https://docs.pytest.org/en/latest/how-to/writing_plugins.html#pip-installable-plugins)).

**Comment by bskinn** [comment]
Yeah, I've never written a plugin system before, so I definitely will be consulting existing best practices. My current plan for the cascade of how a pluggable would be selected is in the numbered list in the original issue comment.

(1) and (2) would just involve passing/storing callable Python objects, simple enough.

For (3) and (4), I figure I would use the same entry-point syntax as `setuptools`: `package.subpackage.subpackage:scorer`, where `scorer` would need to be a suitable callable -- `sphobjinv` would just `from package.subpackage.subpackage import scorer as scoring_func`, and then use `scoring_func` appropriately.

For (5), I would want to use the stricter approach of `entry_points` -- I don't really like the design pattern of casting about the importable namespace for things that look like they have the right format, and then attempting an import of a specific magic function. Too many edge cases, and too much room for footgun.  Also, if we use the [`flake8` style approach to `entry_points`](https://github.com/bskinn/flake8-absolute-import/blob/2c1f0fe7b5b323ef41d365f6ab61a58a41f5ca34/setup.cfg#L46-L48), then there's a lot of flexibility to have multiple scorers installed at the same time. 

```
[options.entry_points]
sphobjinv.scorer = my_scorer_name=my_scorer.scorer
```

I guess that's one of the big questions to decide, though -- will `suggest` only use an alternative plugged-in scorer if explicitly instructed to do so, somehow? Or would there be some heuristic for automatically choosing an available pluggable scorer, if one or more are installed/visible?

**Comment by bskinn** [comment]
Hehe, exactly what you said:

> How would a command-line user select an alternative scoring mechanism? Unless either all detected scorers would be used, or some heuristic would select a scoring mechanism based on the query, then presumably a command line option would name an alternate scoring function, right?

**Comment by bskinn** [comment]
The CLI arg for (3) could line up with the `entry_points` entry... e.g., `sphobjinv suggest --scorer my_scorer_name ...`.

Then, the env variable for (4) would just be something like `SPHOBJINV_SCORER=my_scorer_name`.


Ahh, and (5) wouldn't be a mechanism for *specifying* the scorer to use, but just for *provisioning* a scorer to be available...!

**Comment by bskinn** [comment]
For plugin `suggest` functionality, need to provide hooks for setup functions, for both one-at-a-time and all-at-once scorers, and then also a closeout hook for at least the all-at-once mode. Not immediately obvious why you would need a closeout for the one-at-a-time... Except maybe to dispose resources. Probably best to expose it, JIC.

Should also expose a mechanism to let the plugin define the default suggest threshold for CLI display, because different metrics may tend to lie in different ranges of 0-100.

**Comment by bskinn** [comment]
Will want an API and CLI for listing available scorers.

Will want API and CLI for picking a scorer based on an entry_point spec, as opposed to *only* exposing scorers via name/ID. This should just be a matter of exposing functionality created as part of the `entry_points` integration implementation here.

**Comment by bskinn** [comment]
Sprinkling this in various issues: [`rapidfuzz`](https://pypi.org/project/rapidfuzz/) as a `fuzzywuzzy` alternative.
